plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,466 @@
+      for (String baseVersion : baseVersions) {
+        metadataBuilder.onEnterBaseVersion(baseVersion);
+        try (StorageTx tx = storageFacet.openTx()) {
+          final Iterable<Component> components = tx.findComponents(
+              "group = :groupId and name = :artifactId and attributes.maven2.baseVersion = :baseVersion",
+              ImmutableMap.<String, Object>of(
+                  "groupId", groupId,
+                  "artifactId", artifactId,
+                  "baseVersion", baseVersion
+              ),
+              ImmutableList.of(repository),
+              null // order by
+          );
+          for (Component component : components) {
+            final Iterable<Asset> assets = tx.browseAssets(component);
+            for (Asset asset : assets) {
+              final MavenPath mavenPath = mavenPathParser.parsePath(
+                  asset.formatAttributes().require(StorageFacet.P_PATH, String.class)
+              );
+              if (mavenPath.isSubordinate()) {
+                continue;
+              }
+              metadataBuilder.addArtifactVersion(mavenPath);
+              mayUpdateChecksum(tx, asset, mavenPath, HashType.SHA1);
+              mayUpdateChecksum(tx, asset, mavenPath, HashType.MD5);
+              if (mavenPath.isPom()) {
+                final Document pom = getModel(tx, mavenPath);
+                if (pom != null) {
+                  final String packaging = getChildValue(pom, "packaging", "jar");
+                  log.debug("POM packaging: {}", packaging);
+                  if ("maven-plugin".equals(packaging)) {
+                    metadataBuilder.addPlugin(getPluginPrefix(tx, mavenPath.locateMainArtifact("jar")), artifactId,
+                        getChildValue(pom, "name", null));
+                  }
+                }
+              }
+            }
+          }
+          metadataUpdater.processMetadata(
+              tx,
+              metadataMavenPath(groupId, artifactId, baseVersion),
+              metadataBuilder.onExitBaseVersion()
+          );
+          tx.commit();
+        }
+      }
+      try (StorageTx tx = storageFacet.openTx()) {
+        metadataUpdater.processMetadata(
+            tx,
+            metadataMavenPath(groupId, artifactId, null),
+            metadataBuilder.onExitArtifactId()
+        );
+        tx.commit();
+      }
+    }
+
+    /**
+     * Verifies and may fix/create the broken/non-existent Maven hashes (.sha1/.md5 files).
+     */
+    private void mayUpdateChecksum(final StorageTx tx, final Asset asset, final MavenPath mavenPath,
+                                   final HashType hashType)
+    {
+      final AttributesMap checksums = asset.attributes().child(StorageFacet.P_CHECKSUM);
+      checkState(checksums != null, "checksums");
+      final String assetChecksum = (String) checksums.get(hashType.getHashAlgorithm().name());
+      checkState(!Strings.isNullOrEmpty(assetChecksum), "assetChecksum");
```

Comment:
No, it's opposite: if we encounter a maven asset w/o checksum as defined by format, then we die... but yes, good point what should happen here.